feat(linear_algebra/decomposition): Define module decompositions and prove basic facts

winston-h-zhang opened the pull request:
We define a new structure `decomposition ι R M` as the type of `ι`-indexed decompositions of `M`. We provide a correspondence between internal direct sums and decompositions. Some helpful auxiliary lemmas have been added as well.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
Suggestions or criticisms about how I could do this better is certainly welcome!

- [x] depends on: #8274 
- [x] depends on: #8475

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

winston-h-zhang committed:
initial commit

winston-h-zhang committed:
fix

winston-h-zhang committed:
lint

winston-h-zhang committed:
fix

winston-h-zhang committed:
nolint inhabited_instance

winston-h-zhang committed:
add coe_sum

winston-h-zhang committed:
define decompositions

winston-h-zhang committed:
lint

winston-h-zhang committed:
lint again

winston-h-zhang committed:
equiv between `is_internal` and decomposition

winston-h-zhang committed:
port new stuff

winston-h-zhang committed:
lint

winston-h-zhang committed:
finish moving stuff

winston-h-zhang committed:
missed a file

winston-h-zhang committed:
final pass

winston-h-zhang committed:
doc-string

winston-h-zhang committed:
move file to linear_algebra

eric-wieser commented on src/algebra/module/submodule.lean:
```diff
@@ -183,6 +183,10 @@ by refine {to_fun := coe, ..}; simp [coe_smul]
 
+@[simp] lemma coe_sum {ι : Type*} (x : ι → p) (s : finset ι) :
+  (↑(∑ i in s, x i) : M) = ∑ i in s, ↑(x i) :=
+by convert p.subtype.to_add_monoid_hom.map_sum _ _
```
I'm surprised you need a `convert` here

eric-wieser replied:
Indeed, you don't: extracted to #8246

eric-wieser commented on src/data/dfinsupp.lean:
```diff
@@ -651,6 +651,13 @@ lemma support_subset_iff {s : set ι} {f : Π₀ i, β i} :
 by simp [set.subset_def];
    exact forall_congr (assume i, not_imp_comm)
 
+lemma support_single_eq_zero {i : ι} {b : β i} (hb : b = 0) : (single i b).support = ∅ :=
```
This lemma seems a bit silly to me, since the caller can just `rw h`. I think it would be better as:
```suggestion
lemma support_single_eq_zero {i : ι} : (single i (0 : β i)).support = ∅ :=
```
Which we might already have elsewhere

eric-wieser replied:
Indeed, `by simp [hb]` or `by rw [hb, single_zero, support_zero]` closes the goal, so I don't think this lemma is worth having.

eric-wieser commented on src/data/finsupp/basic.lean:
```diff
@@ -632,6 +632,10 @@ def prod [has_zero M] [comm_monoid N] (f : α →₀ M) (g : α → M → N) : N
 variables [has_zero M] [has_zero M'] [comm_monoid N]
 
+@[to_additive]
+lemma prod_def (f : α →₀ M) (g : α → M → N) :
```
There's no need for this lemma, you can use `rw prod` anywhere `rw prod_def` would work.

eric-wieser commented on src/data/finsupp/basic.lean:
```diff
@@ -1064,6 +1068,23 @@ lemma prod_add_index' [add_comm_monoid M] [comm_monoid N] {f g : α →₀ M}
     f.prod (λ a b, h a (multiplicative.of_add b)) * g.prod (λ a b, h a (multiplicative.of_add b)) :=
 prod_add_index (λ a, (h a).map_one) (λ a, (h a).map_mul)
 
+lemma sum_add_add
```
This looks like a special case of [`finsupp.sum_add_index`](https://leanprover-community.github.io/mathlib_docs/data/finsupp/basic.html#finsupp.sum_add_index)  to me, with `h = id`

eric-wieser commented on src/data/fintype/basic.lean:
```diff
@@ -1636,3 +1636,12 @@ variables [fintype α] [decidable_eq α]
+namespace fintype
+variables [fintype α] [complete_lattice β]
+
+lemma sup_eq_supr (f : α → β) : finset.univ.sup f = supr f :=
```
```suggestion
/-- A special case of `finset.sup_eq_supr` with a better RHS. -/
lemma sup_eq_supr (f : α → β) : finset.univ.sup f = supr f :=
```
I'd also be tempted to rename this to `finset.sup_univ_eq_supr` to match the other lemma.

eric-wieser replied:
I've PR'd this as #8397, along with the `inf` version.

eric-wieser commented on src/linear_algebra/decomposition.lean:
```diff
@@ -0,0 +1,263 @@
+
+open_locale direct_sum classical big_operators
+open classical linear_map dfinsupp direct_sum
+noncomputable theory
```
I'd be interested to know exactly which bits of this are noncomputable.

My hunch right now is that this is not a good choice of definition of `decomposition`; I think the decomposition should just hold the actual equiv, and you should have a `noncomputablr def decomposition.of_independent` that constructs one from the fields you currently have.

winston-h-zhang replied:
Are there code/formalization specific benefits to holding to actual equiv? I'm not attached to my definition; they all look equivalent to me.

```lean
def decomposition (ι : Type u) (R : Type v) (M : Type w)
  [semiring R] [add_comm_monoid M] [module R M] := 
subtype (λ (p : ι → submodule R M), submodule_is_internal p)


structure decomposition' (ι : Type u) (R : Type v) (M : Type w)
  [semiring R] [add_comm_monoid M] [module R M] := 
(factors : ι → submodule R M)
(to_equiv : (⨁ i, factors i) ≃ₗ[R] M)
```

These were the two other definitions I was thinking about, but they honestly look exactly the same :P

winston-h-zhang replied:
To be clear by "hold the actual equiv" I assume you mean something like the second one?

eric-wieser replied:
Yes, although I guess you don't want to hold the equiv, rather just the inverse and the proof it is an inverse of the function we care about.

The advantage of doing that is you can compute with decompositions.

eric-wieser replied:
I can reply with a mwe on Monday, but can't really do so from mobile.

eric-wieser commented on src/order/complete_lattice.lean:
```diff
@@ -473,6 +473,19 @@ le_antisymm
 
+-- Unfortunately `ι : Sort u_4` doesn't work with `set ι`
+theorem bsupr_eq_supr {ι : Type*} {s : ι → α}
+  {p : set ι} : (⨆ i (H : p i), s i) = ⨆ (i : p), s i :=
```
```suggestion
  {p : set ι} : (⨆ i (H : i ∈ p), s i) = ⨆ (i : p), s i :=
```
or
```suggestion
  {p : ι → Prop} : (⨆ i (H : p i), s i) = ⨆ (i : subtype p), s i :=
```
Don't ever use `s x` to mean `s ∈ x`, it's not the canonical spelling and leaves the caller a mess to clean up

eric-wieser replied:
We might already have this lemma in one of those spellings

eric-wieser replied:
Yes, this is just `supr_subtype`

eric-wieser reviewed (commented):
Thanks for taking this on! 🎉

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -1215,6 +1230,101 @@ begin
+  { rw ← hsum, exact sum_mem_bsupr (λ i hi, hv i), },
+  exact hle key,
+end
+
```
I think it might be worth splitting these two proofs (and anything you added in order to prove them) into their own smaller PR, and asking zulip for help golfing them. This PR is great, but it's also large, and if you split it up you can get more focussed feedback on each part.

eric-wieser committed:
feat({data,linear_algebra}/{finsupp,dfinsupp}): add `{add_submonoid,submodule}.[d]finsupp_sum_mem`

These lemmas are trivial consequences of the finset lemmas, but having them avoids having to unfold `[d]finsupp.sum`.

`dfinsupp_sum_add_hom_mem` is particularly useful because this one has some messy decidability arguments to eliminate.

eric-wieser committed:
fix

eric-wieser committed:
fix

eric-wieser committed:
feat(algebra/direct_sum): the submodules of an internal direct sum satisfy `supr A = ⊤`

eric-wieser commented on src/linear_algebra/decomposition.lean:
```diff
@@ -0,0 +1,263 @@
+    rw [submodule_is_internal.to_equiv_symm_single_apply p hp i x hi, single_eq_of_ne], refl,
+    exact h
+  end,
+  factors_supr := begin
```
I've PR'd a shorter proof of this in #8274, which also eliminates the need for `mem_supr`

eric-wieser replied:
I went ahead and pushed a commit to this PR that uses the new lemma.

eric-wieser committed:
Merge branch 'master' into eric-wieser/direct_sum.is_internal.supr_eq_top

winston-h-zhang committed:
Merge branch 'eric-wieser/direct_sum.is_internal.supr_eq_top' of https://github.com/leanprover-community/mathlib into module-decompositions

eric-wieser commented on src/linear_algebra/decomposition.lean:
```diff
@@ -0,0 +1,263 @@
+        mk_eq_zero, not_imp_self, ite_eq_right_iff, hzero], },
+    rw [submodule_is_internal.to_equiv_symm_single_apply p hp i x hi, single_eq_of_ne], refl,
+    exact h
+  end,
```
I think you should extract this to a standalone lemma like
```lean
lemma submodule_is_internal.independent {p : ι → submodule R M} (hp : direct_sum.submodule_is_internal p) : complete_lattice.independent A :=
sorry -- this `begin` / `end` block
```

winston-h-zhang committed:
Merge branch 'master' into module-decompositions

eric-wieser committed:
Merge branch 'master' into module-decompositions

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1081,6 +1081,17 @@ begin
+  rw [mul_comm, ← prod_insert, insert_erase],
+  exacts [h, not_mem_erase a s],
+end
+
```
PR'd separately as #8449

eric-wieser committed:
fix bad merge

eric-wieser committed:
Use existing proof

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into module-decompositions

eric-wieser commented on src/order/compactly_generated.lean:
```diff
@@ -470,8 +473,23 @@ theorem is_complemented_of_Sup_atoms_eq_top (h : Sup {a : α | is_atom a} = ⊤)
+  obtain ⟨s, s_ind, s_top, s_atoms⟩ := exists_set_independent_Sup_atoms_eq_compl h (⊥ : α),
+  refine ⟨s, s_ind, eq_top_of_is_compl_bot (is_compl.symm s_top), s_atoms⟩,
+end
+
```
@acxxa, do you use the new lemmas in this file in the rest of the PR? If not, can you split this out into its own PR?

winston-h-zhang replied:
I've split it out in #8475

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into module-decompositions

winston-h-zhang committed:
oh noes

winston-h-zhang committed:
Merge branch 'module-decompositions' of https://github.com/leanprover-community/mathlib into module-decompositions

winston-h-zhang committed:
fix and lint

winston-h-zhang committed:
fix

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into module-decompositions

I haven't attempted to build this locally, I will let CI decide its fate.

eric-wieser committed:
remove duplicate or unnecessary lemmas

These lemmas already exists in inconsequentially different ways, it's better to just reuse the ones we have.

This golfs some proofs too.

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -1131,6 +1131,21 @@ begin
+  rw finset.mem_erase at hc,
+  have almost : p' c hc.1 ≤ ⨆ i (H : i ≠ j), p' i H := le_bsupr _ _,
+  exact almost (hx c),
+end
```
Not if you reuse what we already have:
```suggestion
lemma sum_erase_mem_bsupr {ι : Type*} [fintype ι] [decidable_eq ι]
  (x : ι → M) {p : ι → submodule R M} (j : ι) (hx : ∀ i, x i ∈ p i) :
  ∑ (i : ι) in finset.univ.erase j, (x i) ∈ ⨆ (k : ι) (H : k ≠ j), p k :=
begin
  convert sum_mem_bsupr (λ i hi, hx i),
  simp,
end
```

eric-wieser committed:
golfs

eric-wieser committed:
golf

eric-wieser commented on src/linear_algebra/basic.lean:
```diff
@@ -1355,6 +1363,101 @@ begin
+-- this is **definitely** going to be atrocious
+lemma mem_bsupr {p : ι → Prop} (f : ι → submodule R M) (x : M) :
+  x ∈ (⨆ i (H : p i), f i) ↔
+  ∃ v : ι →₀ M, (∀ i, v i ∈ f i) ∧ ∑ i in v.support, v i = x ∧ (∀ i, ¬ p i → v i = 0) :=
```
I've PR'd an alternate version of this as #9202

eric-wieser committed:
Merge remote-tracking branch 'origin/master' into module-decompositions

This removes some proofs which now already exist more generally

eric-wieser committed:
Golf some proofs

eric-wieser commented on src/linear_algebra/decomposition.lean:
```diff
@@ -0,0 +1,170 @@
+which phased out of the `set` design because of problems with it.
+-/
+@[nolint has_inhabited_instance]
+structure decomposition (ι : Type u) (R : Type v) (M : Type w)
```
This bears a resemblance to #10115. Both approaches have their merits, but I hope we can get the best of both.

YaelDillies committed:
Merge remote-tracking branch 'origin/master' into module-decompositions

YaelDillies committed:
typo

YaelDillies committed:
fix
